Firstly I load the packages I need.

In [93]:
import io
import requests
import pandas as pd
import numpy as np

Secondly I use the pandas .read_html method to scrape the table with the neighbourhoods in Toronto from the Wikipedia page. I specified by using [0] for the particular table I want since the page contains more than one table. 

In [94]:
Toronto = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header=0,
                     )
Toronto[0]

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
35       M7C      Not assigned   
36       M8C      Not assigned   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M2E      Not assigned   
45       M3E      Not assigned   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
49       M7E      Not assigned   
50       M8E      Not assigned   
51       M9E      Not assigned   
52       M1G       Scarborough   
53       M2G      Not assigned   
54       M3G      Not assigned   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
58       M7G      Not assigned   
59       M8G      Not assigned   
60       M9G      Not assigned   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
72       M7H      Not assigned   
73       M8H      Not assigned   
74       M9H      Not assigned   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
87       M7J      Not assigned   
88       M8J      Not assigned   
89       M9J      Not assigned   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M7K      Not assigned   
104      M8K      Not assigned   
105      M9K      Not assigned   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North 

The scraped table contains 287 rows and 3 columns. Then I convert the table into a pandas dataframe called df.

In [95]:
df = pd.DataFrame(Toronto[0])
df

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
35       M7C      Not assigned   
36       M8C      Not assigned   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M2E      Not assigned   
45       M3E      Not assigned   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
49       M7E      Not assigned   
50       M8E      Not assigned   
51       M9E      Not assigned   
52       M1G       Scarborough   
53       M2G      Not assigned   
54       M3G      Not assigned   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
58       M7G      Not assigned   
59       M8G      Not assigned   
60       M9G      Not assigned   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
72       M7H      Not assigned   
73       M8H      Not assigned   
74       M9H      Not assigned   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
87       M7J      Not assigned   
88       M8J      Not assigned   
89       M9J      Not assigned   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M7K      Not assigned   
104      M8K      Not assigned   
105      M9K      Not assigned   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North 

I then delete all cells with a Borough that is Not assigned so that I only process the cells that have an assigned borough.

In [96]:
df = df[df.Borough != 'Not assigned']
df

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
146      M6N              York   
149      M9N              York   
150      M1P       Scarbor

This reduces my dataframe from 287 rows to 210 rows, meaning that 77 cells had a Borough that was Not assigned. 

I then combine into one row all Neighbourhoods with the same Postcode, with the Neighbourhoods separated with a comma. To do that, firstly I use the .groupby method on condition of the Neighbourhoods having the same Postcode and then use the .apply method to execute the grouping and the .join method to separate the Neighbourhoods with a coma and save the result in a new dataframe called df1.

In [97]:
df1 = df.groupby('Postcode')['Neighbourhood'].apply(','.join).reset_index()
df1

Postcode                                      Neighbourhood
0        M1B                                      Rouge,Malvern
1        M1C               Highland Creek,Rouge Hill,Port Union
2        M1E                    Guildwood,Morningside,West Hill
3        M1G                                             Woburn
4        M1H                                          Cedarbrae
5        M1J                                Scarborough Village
6        M1K          East Birchmount Park,Ionview,Kennedy Park
7        M1L                      Clairlea,Golden Mile,Oakridge
8        M1M      Cliffcrest,Cliffside,Scarborough Village West
9        M1N                         Birch Cliff,Cliffside West
10       M1P  Dorset Park,Scarborough Town Centre,Wexford He...
11       M1R                                   Maryvale,Wexford
12       M1S                                          Agincourt
13       M1T              Clarks Corners,Sullivan,Tam O'Shanter
14       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...
15       M1W                                    L'Amoreaux West
16       M1X                                        Upper Rouge
17       M2H                                  Hillcrest Village
18       M2J                         Fairview,Henry Farm,Oriole
19       M2K                                    Bayview Village
20       M2L                            Silver Hills,York Mills
21       M2M                             Newtonbrook,Willowdale
22       M2N                                   Willowdale South
23       M2P                                    York Mills West
24       M2R                                    Willowdale West
25       M3A                                          Parkwoods
26       M3B                                    Don Mills North
27       M3C                    Flemingdon Park,Don Mills South
28       M3H      Bathurst Manor,Downsview North,Wilson Heights
29       M3J                     Northwood Park,York University
30       M3K                         CFB Toronto,Downsview East
31       M3L                                     Downsview West
32       M3M                                  Downsview Central
33       M3N                                Downsview Northwest
34       M4A                                   Victoria Village
35       M4B                     Woodbine Gardens,Parkview Hill
36       M4C                                   Woodbine Heights
37       M4E                                        The Beaches
38       M4G                                            Leaside
39       M4H                                   Thorncliffe Park
40       M4J                                       East Toronto
41       M4K                        The Danforth West,Riverdale
42       M4L                      The Beaches West,India Bazaar
43       M4M                                    Studio District
44       M4N                                      Lawrence Park
45       M4P                                   Davisville North
46       M4R                                 North Toronto West
47       M4S                                         Davisville
48       M4T                         Moore Park,Summerhill East
49       M4V  Deer Park,Forest Hill SE,Rathnelly,South Hill,...
50       M4W                                           Rosedale
51       M4X                         Cabbagetown,St. James Town
52       M4Y                               Church and Wellesley
53       M5A                                       Harbourfront
54       M5B                            Ryerson,Garden District
55       M5C                                     St. James Town
56       M5E                                        Berczy Park
57       M5G                                 Central Bay Street
58       M5H                             Adelaide,King,Richmond
59       M5J    Harbourfront East,Toronto Islands,Union Station
60       M5K            Design Exchange,Toronto Dominion Centre
61       M5L                      Co

The returned output above excludes the Borough column. I use the pandas .merge method to combine the Borough column from the df dataframe and the df1 dataframe on condition that they're merged on matching the Borough to the Neighbourhood based on the Postcode and I save the results in a new dataframe called df2.

In [98]:
df2 = pd.merge(df1, df[["Postcode", "Borough"]], on="Postcode")
df2

Postcode                                      Neighbourhood  \
0        M1B                                      Rouge,Malvern   
1        M1B                                      Rouge,Malvern   
2        M1C               Highland Creek,Rouge Hill,Port Union   
3        M1C               Highland Creek,Rouge Hill,Port Union   
4        M1C               Highland Creek,Rouge Hill,Port Union   
5        M1E                    Guildwood,Morningside,West Hill   
6        M1E                    Guildwood,Morningside,West Hill   
7        M1E                    Guildwood,Morningside,West Hill   
8        M1G                                             Woburn   
9        M1H                                          Cedarbrae   
10       M1J                                Scarborough Village   
11       M1K          East Birchmount Park,Ionview,Kennedy Park   
12       M1K          East Birchmount Park,Ionview,Kennedy Park   
13       M1K          East Birchmount Park,Ionview,Kennedy Park   
14       M1L                      Clairlea,Golden Mile,Oakridge   
15       M1L                      Clairlea,Golden Mile,Oakridge   
16       M1L                      Clairlea,Golden Mile,Oakridge   
17       M1M      Cliffcrest,Cliffside,Scarborough Village West   
18       M1M      Cliffcrest,Cliffside,Scarborough Village West   
19       M1M      Cliffcrest,Cliffside,Scarborough Village West   
20       M1N                         Birch Cliff,Cliffside West   
21       M1N                         Birch Cliff,Cliffside West   
22       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   
23       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   
24       M1P  Dorset Park,Scarborough Town Centre,Wexford He...   
25       M1R                                   Maryvale,Wexford   
26       M1R                                   Maryvale,Wexford   
27       M1S                                          Agincourt   
28       M1T              Clarks Corners,Sullivan,Tam O'Shanter   
29       M1T              Clarks Corners,Sullivan,Tam O'Shanter   
30       M1T              Clarks Corners,Sullivan,Tam O'Shanter   
31       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...   
32       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...   
33       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...   
34       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...   
35       M1W                                    L'Amoreaux West   
36       M1X                                        Upper Rouge   
37       M2H                                  Hillcrest Village   
38       M2J                         Fairview,Henry Farm,Oriole   
39       M2J                         Fairview,Henry Farm,Oriole   
40       M2J                         Fairview,Henry Farm,Oriole   
41       M2K                                    Bayview Village   
42       M2L                            Silver Hills,York Mills   
43       M2L                            Silver Hills,York Mills   
44       M2M                             Newtonbrook,Willowdale   
45       M2M                             Newtonbrook,Willowdale   
46       M2N                                   Willowdale South   
47       M2P                                    York Mills West   
48       M2R                                    Willowdale West   
49       M3A                                          Parkwoods   
50       M3B                                    Don Mills North   
51       M3C                    Flemingdon Park,Don Mills South   
52       M3C                    Flemingdon Park,Don Mills South   
53       M3H      Bathurst Manor,Downsview North,Wilson Heights   
54       M3H      Bathurst Manor,Downsview North,Wilson Heights   
55       M3H      Bathurst Manor,Downsview North,Wilson Heights   
56       M3J                     Northwood Park,York University   
57       M3J                     Northwood Park,York University   
58       M3K                         CFB Toronto,Do

I then rearrange the order of my columns to start with Postcode, followed by Borough and thirdly Neighbourhood.

In [99]:
df2 = df2[['Postcode', 'Borough', 'Neighbourhood']]
df2

Postcode           Borough  \
0        M1B       Scarborough   
1        M1B       Scarborough   
2        M1C       Scarborough   
3        M1C       Scarborough   
4        M1C       Scarborough   
5        M1E       Scarborough   
6        M1E       Scarborough   
7        M1E       Scarborough   
8        M1G       Scarborough   
9        M1H       Scarborough   
10       M1J       Scarborough   
11       M1K       Scarborough   
12       M1K       Scarborough   
13       M1K       Scarborough   
14       M1L       Scarborough   
15       M1L       Scarborough   
16       M1L       Scarborough   
17       M1M       Scarborough   
18       M1M       Scarborough   
19       M1M       Scarborough   
20       M1N       Scarborough   
21       M1N       Scarborough   
22       M1P       Scarborough   
23       M1P       Scarborough   
24       M1P       Scarborough   
25       M1R       Scarborough   
26       M1R       Scarborough   
27       M1S       Scarborough   
28       M1T       Scarborough   
29       M1T       Scarborough   
30       M1T       Scarborough   
31       M1V       Scarborough   
32       M1V       Scarborough   
33       M1V       Scarborough   
34       M1V       Scarborough   
35       M1W       Scarborough   
36       M1X       Scarborough   
37       M2H        North York   
38       M2J        North York   
39       M2J        North York   
40       M2J        North York   
41       M2K        North York   
42       M2L        North York   
43       M2L        North York   
44       M2M        North York   
45       M2M        North York   
46       M2N        North York   
47       M2P        North York   
48       M2R        North York   
49       M3A        North York   
50       M3B        North York   
51       M3C        North York   
52       M3C        North York   
53       M3H        North York   
54       M3H        North York   
55       M3H        North York   
56       M3J        North York   
57       M3J        North York   
58       M3K        North York   
59       M3K        North York   
60       M3L        North York   
61       M3M        North York   
62       M3N        North York   
63       M4A        North York   
64       M4B         East York   
65       M4B         East York   
66       M4C         East York   
67       M4E      East Toronto   
68       M4G         East York   
69       M4H         East York   
70       M4J         East York   
71       M4K      East Toronto   
72       M4K      East Toronto   
73       M4L      East Toronto   
74       M4L      East Toronto   
75       M4M      East Toronto   
76       M4N   Central Toronto   
77       M4P   Central Toronto   
78       M4R   Central Toronto   
79       M4S   Central Toronto   
80       M4T   Central Toronto   
81       M4T   Central Toronto   
82       M4V   Central Toronto   
83       M4V   Central Toronto   
84       M4V   Central Toronto   
85       M4V   Central Toronto   
86       M4V   Central Toronto   
87       M4W  Downtown Toronto   
88       M4X  Downtown Toronto   
89       M4X  Downtown Toronto   
90       M4Y  Downtown Toronto   
91       M5A  Downtown Toronto   
92       M5B  Downtown Toronto   
93       M5B  Downtown Toronto   
94       M5C  Downtown Toronto   
95       M5E  Downtown Toronto   
96       M5G  Downtown Toronto   
97       M5H  Downtown Toronto   
98       M5H  Downtown Toronto   
99       M5H  Downtown Toronto   
100      M5J  Downtown Toronto   
101      M5J  Downtown Toronto   
102      M5J  Downtown Toronto   
103      M5K  Downtown Toronto   
104      M5K  Downtown Toronto   
105      M5L  Downtown Toronto   
106      M5L  Downtown Toronto   
107      M5M        North York   
108      M5M        North York   
109      M5N   Central Toronto   
110      M5P   Central Toronto   
111      M5P   Central Toronto   
112      M5R   Central Toronto   
113      M5R   Central Toronto   
114      M5R   Central Toronto   
115      M5S  Downtown Toronto   
116      M5S  Downtown Tor

I then drop all resulting duplicates based on the same Postcode and save the results in a new dataframe called df3.

In [100]:
df3 = df2.drop_duplicates(subset="Postcode")
df3

Postcode           Borough  \
0        M1B       Scarborough   
2        M1C       Scarborough   
5        M1E       Scarborough   
8        M1G       Scarborough   
9        M1H       Scarborough   
10       M1J       Scarborough   
11       M1K       Scarborough   
14       M1L       Scarborough   
17       M1M       Scarborough   
20       M1N       Scarborough   
22       M1P       Scarborough   
25       M1R       Scarborough   
27       M1S       Scarborough   
28       M1T       Scarborough   
31       M1V       Scarborough   
35       M1W       Scarborough   
36       M1X       Scarborough   
37       M2H        North York   
38       M2J        North York   
41       M2K        North York   
42       M2L        North York   
44       M2M        North York   
46       M2N        North York   
47       M2P        North York   
48       M2R        North York   
49       M3A        North York   
50       M3B        North York   
51       M3C        North York   
53       M3H        North York   
56       M3J        North York   
58       M3K        North York   
60       M3L        North York   
61       M3M        North York   
62       M3N        North York   
63       M4A        North York   
64       M4B         East York   
66       M4C         East York   
67       M4E      East Toronto   
68       M4G         East York   
69       M4H         East York   
70       M4J         East York   
71       M4K      East Toronto   
73       M4L      East Toronto   
75       M4M      East Toronto   
76       M4N   Central Toronto   
77       M4P   Central Toronto   
78       M4R   Central Toronto   
79       M4S   Central Toronto   
80       M4T   Central Toronto   
82       M4V   Central Toronto   
87       M4W  Downtown Toronto   
88       M4X  Downtown Toronto   
90       M4Y  Downtown Toronto   
91       M5A  Downtown Toronto   
92       M5B  Downtown Toronto   
94       M5C  Downtown Toronto   
95       M5E  Downtown Toronto   
96       M5G  Downtown Toronto   
97       M5H  Downtown Toronto   
100      M5J  Downtown Toronto   
103      M5K  Downtown Toronto   
105      M5L  Downtown Toronto   
107      M5M        North York   
109      M5N   Central Toronto   
110      M5P   Central Toronto   
112      M5R   Central Toronto   
115      M5S  Downtown Toronto   
117      M5T  Downtown Toronto   
120      M5V  Downtown Toronto   
127      M5W  Downtown Toronto   
128      M5X  Downtown Toronto   
130      M6A        North York   
132      M6B        North York   
133      M6C              York   
134      M6E              York   
135      M6G  Downtown Toronto   
136      M6H      West Toronto   
138      M6J      West Toronto   
140      M6K      West Toronto   
143      M6L        North York   
146      M6M              York   
150      M6N              York   
152      M6P      West Toronto   
154      M6R      West Toronto   
156      M6S      West Toronto   
158      M7A      Queen's Park   
159      M7R       Mississauga   
160      M7Y      East Toronto   
161      M8V         Etobicoke   
164      M8W         Etobicoke   
166      M8X         Etobicoke   
169      M8Y         Etobicoke   
177      M8Z         Etobicoke   
182      M9A      Queen's Park   
183      M9B         Etobicoke   
188      M9C         Etobicoke   
192      M9L        North York   
193      M9M        North York   
195      M9N              York   
196      M9P         Etobicoke   
197      M9R         Etobicoke   
201      M9V         Etobicoke   
209      M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
2                 Highland Creek,Rouge Hill,Port Union  
5                      Guildwood,Morningside,West Hill  
8                                               Woburn  
9                                            Cedarbrae  
10                                 Scarborough Village  
11           East Birchmount Park,Ionview,Kennedy Park  
14         

The code below is to find out how many Neighbourhoods are Not assigned. The result is just one, with a Borough named Queen's Park.

In [101]:
print(df3.loc[df3['Neighbourhood'] == 'Not assigned'])

    Postcode       Borough Neighbourhood
158      M7A  Queen's Park  Not assigned


I then replace the Not assigned Neighbourhood with its Borough name by using the .loc method.

In [102]:
df3.loc[df3['Neighbourhood'] == 'Not assigned'] = "Queen's Park"
df3

/usr/local/lib/python3.5/dist-packages/pandas/core/indexing.py:189: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
/usr/local/lib/python3.5/dist-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Postcode           Borough  \
0             M1B       Scarborough   
2             M1C       Scarborough   
5             M1E       Scarborough   
8             M1G       Scarborough   
9             M1H       Scarborough   
10            M1J       Scarborough   
11            M1K       Scarborough   
14            M1L       Scarborough   
17            M1M       Scarborough   
20            M1N       Scarborough   
22            M1P       Scarborough   
25            M1R       Scarborough   
27            M1S       Scarborough   
28            M1T       Scarborough   
31            M1V       Scarborough   
35            M1W       Scarborough   
36            M1X       Scarborough   
37            M2H        North York   
38            M2J        North York   
41            M2K        North York   
42            M2L        North York   
44            M2M        North York   
46            M2N        North York   
47            M2P        North York   
48            M2R        North York   
49            M3A        North York   
50            M3B        North York   
51            M3C        North York   
53            M3H        North York   
56            M3J        North York   
58            M3K        North York   
60            M3L        North York   
61            M3M        North York   
62            M3N        North York   
63            M4A        North York   
64            M4B         East York   
66            M4C         East York   
67            M4E      East Toronto   
68            M4G         East York   
69            M4H         East York   
70            M4J         East York   
71            M4K      East Toronto   
73            M4L      East Toronto   
75            M4M      East Toronto   
76            M4N   Central Toronto   
77            M4P   Central Toronto   
78            M4R   Central Toronto   
79            M4S   Central Toronto   
80            M4T   Central Toronto   
82            M4V   Central Toronto   
87            M4W  Downtown Toronto   
88            M4X  Downtown Toronto   
90            M4Y  Downtown Toronto   
91            M5A  Downtown Toronto   
92            M5B  Downtown Toronto   
94            M5C  Downtown Toronto   
95            M5E  Downtown Toronto   
96            M5G  Downtown Toronto   
97            M5H  Downtown Toronto   
100           M5J  Downtown Toronto   
103           M5K  Downtown Toronto   
105           M5L  Downtown Toronto   
107           M5M        North York   
109           M5N   Central Toronto   
110           M5P   Central Toronto   
112           M5R   Central Toronto   
115           M5S  Downtown Toronto   
117           M5T  Downtown Toronto   
120           M5V  Downtown Toronto   
127           M5W  Downtown Toronto   
128           M5X  Downtown Toronto   
130           M6A        North York   
132           M6B        North York   
133           M6C              York   
134           M6E              York   
135           M6G  Downtown Toronto   
136           M6H      West Toronto   
138           M6J      West Toronto   
140           M6K      West Toronto   
143           M6L        North York   
146           M6M              York   
150           M6N              York   
152           M6P      West Toronto   
154           M6R      West Toronto   
156           M6S      West Toronto   
158  Queen's Park      Queen's Park   
159           M7R       Mississauga   
160           M7Y      East Toronto   
161           M8V         Etobicoke   
164           M8W         Etobicoke   
166           M8X         Etobicoke   
169           M8Y         Etobicoke   
177           M8Z         Etobicoke   
182           M9A      Queen's Park   
183           M9B         Etobicoke   
188           M9C         Etobicoke   
192           M9L        North York   
193           M9M        North York   
195           M9N              York   
196           M9P         Etobicoke   
197           M9R         Etobicoke   
201           M9V         Etobi

To ascertain the number of rows and columns in my final df3 dataframe, I use the .shape method and this shows that my final dataframe has 103 rows and 3 columns.

In [103]:
df3.shape

(103, 3)

First I use the pandas .read_csv method to read the cvs file containing the Latitudes and Longitudes of the Toronto Neighbourhoods based on Postcodes and I save the csv data in a pandas dataframe called Coords.

In [104]:
Coords = pd.read_csv("https://cocl.us/Geospatial_data") 
Coords

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

I then rename the Postal Code column in the Coords dataframe to Postcode like the df3 dataframe to make it easier for me to merge the 2 dataframes together.

In [105]:
Coords.rename(columns={'Postal Code':'Postcode'}, inplace=True)

print(Coords.columns)

Index(['Postcode', 'Latitude', 'Longitude'], dtype='object')


Using the pandas .merge method to combine the two dataframes (df3 and Coords) together on condition that they're merged on the Postcode being the same and I save the results in a new dataframe called df4.

In [106]:
df4 = pd.merge(df3, Coords[["Postcode", "Latitude", "Longitude"]], on="Postcode")
df4

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7R       Mississauga   
86       M7Y      East Toronto   
87       M8V         Etobicoke   
88       M8W         Etobicoke   
89       M8X         Etobicoke   
90       M8Y         Etobicoke   
91       M8Z         Etobicoke   
92       M9A      Queen's Park   
93       M9B         Etobicoke   
94       M9C         Etobicoke   
95       M9L        North York   
96       M9M        North York   
97       M9N              York   
98       M9P         Etobicoke   
99       M9R         Etobicoke   
100      M9V         Etobicoke   
101      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                          

The resulting dataframe after the merge has one less row, meaning that one row was dropped maybe due to a mismatch.

## Exploring and clustering the neighbourhoods in Toronto

First I get all the dependencies that I will need, that I hadn't got as yet.

In [107]:
import sys
!{sys.executable} -m pip install geopy

In [108]:
import sys
!{sys.executable} -m pip install folium

In [109]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Checking the number of Boroughs and Neighbourhoods in the dataset.

In [110]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df4['Borough'].unique()),
        df4.shape[0]
    )
)

The dataframe has 11 boroughs and 102 neighborhoods.


Firstly, in order to define an instance of the geocoder, we need to define a user_agent. I will name my user agent toronto_explorer, as shown below.

In [111]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### I then create a map of Toronto with neighborhoods superimposed on top.

In [112]:
# First I create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# then I add markers to map
for lat, lng, Borough, Neighbourhood in zip(df4['Latitude'], df4['Longitude'], df4['Borough'], df4['Neighbourhood']):
    label = '{}, {}'.format(Neighbourhood, Borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

The map shows the neighbourhoods of Toronto superimposed on the map of Toronto.

I then slice the original dataframe (df4) and create a new dataframe of the neighbourhoods in North York Borough and I save the data in a new datafame called northyork_data.

In [113]:
northyork_data = df4[df4['Borough'] == 'North York'].reset_index(drop=True)
northyork_data.head()

Postcode     Borough               Neighbourhood   Latitude  Longitude
0      M2H  North York           Hillcrest Village  43.803762 -79.363452
1      M2J  North York  Fairview,Henry Farm,Oriole  43.778517 -79.346556
2      M2K  North York             Bayview Village  43.786947 -79.385975
3      M2L  North York     Silver Hills,York Mills  43.757490 -79.374714
4      M2M  North York      Newtonbrook,Willowdale  43.789053 -79.408493

Then I get the coordinates of North York.

In [114]:
address = 'North York, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of North York are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of North York are 43.7708175, -79.4132998.


I then visualise North York with the neighbourhoods in it, as shown below.

In [115]:
# create map of North York using latitude and longitude values
map_northyork = folium.Map(location=[latitude, longitude], zoom_start=11)

# I then add markers to map
for lat, lng, label in zip(northyork_data['Latitude'], northyork_data['Longitude'], northyork_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='purple',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_northyork)  
    
map_northyork

#### Defining Foursquare Credentials and Version

In [116]:
CLIENT_ID = 'DK0DZVYAYUF2IWJNSSGAFNWQNELVW13HV2OJOMBFB04RXRIL' # your Foursquare ID
CLIENT_SECRET = '0DOANYV3WCUVAFTLL5K5WYC2CG0F2L4MQMKBRDW2V3EWTDRM' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: DK0DZVYAYUF2IWJNSSGAFNWQNELVW13HV2OJOMBFB04RXRIL
CLIENT_SECRET:0DOANYV3WCUVAFTLL5K5WYC2CG0F2L4MQMKBRDW2V3EWTDRM


## 1. Exploring Neighborhoods in North York

In [117]:
radius = 500
LIMIT = 100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id=DK0DZVYAYUF2IWJNSSGAFNWQNELVW13HV2OJOMBFB04RXRIL&client_secret=0DOANYV3WCUVAFTLL5K5WYC2CG0F2L4MQMKBRDW2V3EWTDRM&v=20180605&ll=43.7708175,-79.4132998&radius=500&limit=30'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

 The code below runs the above function on each neighbourhood and creates a new dataframe called northyork_venues.

In [118]:
northyork_venues = getNearbyVenues(names=northyork_data['Neighbourhood'],
                                   latitudes=northyork_data['Latitude'],
                                   longitudes=northyork_data['Longitude']
                                  )



Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Bedford Park,Lawrence Manor East
Lawrence Heights,Lawrence Manor
Glencairn
Downsview,North Park,Upwood Park
Humber Summit
Emery,Humberlea


##### I then check the size of the resulting dataframe.

In [119]:
print(northyork_venues.shape)
northyork_venues.head()

(720, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Hillcrest Village               43.803762               -79.363452   
1  Hillcrest Village               43.803762               -79.363452   
2  Hillcrest Village               43.803762               -79.363452   
3  Hillcrest Village               43.803762               -79.363452   
4  Hillcrest Village               43.803762               -79.363452   

                           Venue  Venue Latitude  Venue Longitude  \
0             The Captain's Boil       43.773255       -79.413805   
1             Aroma Espresso Bar       43.769449       -79.413081   
2                        Loblaws       43.768722       -79.412101   
3  Cineplex Cinemas Empress Walk       43.768625       -79.412613   
4                  Konjiki Ramen       43.766998       -79.412222   

       Venue Category  
0  Seafood Restaurant  
1                Café  
2       Grocery Store  
3       Movie Theater  
4    Ramen Restaurant

The code below checks how many venues were returned for each neighbourhood.

In [120]:
northyork_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                           
Bathurst Manor,Downsview North,Wilson Heights                      30   
Bayview Village                                                    30   
Bedford Park,Lawrence Manor East                                   30   
CFB Toronto,Downsview East                                         30   
Don Mills North                                                    30   
Downsview Central                                                  30   
Downsview Northwest                                                30   
Downsview West                                                     30   
Downsview,North Park,Upwood Park                                   30   
Emery,Humberlea                                                    30   
Fairview,Henry Farm,Oriole                                         30   
Flemingdon Park,Don Mills South                                    30   
Glencairn                                                          30   
Hillcrest Village                                                  30   
Humber Summit                                                      30   
Lawrence Heights,Lawrence Manor                                    30   
Newtonbrook,Willowdale                                             30   
Northwood Park,York University                                     30   
Parkwoods                                                          30   
Silver Hills,York Mills                                            30   
Victoria Village                                                   30   
Willowdale South                                                   30   
Willowdale West                                                    30   
York Mills West                                                    30   

                                               Neighbourhood Longitude  Venue  \
Neighbourhood                                                                   
Bathurst Manor,Downsview North,Wilson Heights                       30     30   
Bayview Village                                                     30     30   
Bedford Park,Lawrence Manor East                                    30     30   
CFB Toronto,Downsview East                                          30     30   
Don Mills North                                                     30     30   
Downsview Central                                                   30     30   
Downsview Northwest                                                 30     30   
Downsview West                                                      30     30   
Downsview,North Park,Upwood Park                                    30     30   
Emery,Humberlea                                                     30     30   
Fairview,Henry Farm,Oriole                                          30     30   
Flemingdon Park,Don Mills South                                     30     30   
Glencairn                                                           30     30   
Hillcrest Village                                                   30     30   
Humber Summit                                                       30     30   
Lawrence Heights,Lawrence Manor                                     30     30   
Newtonbrook,Willowdale                                              30     30   
Northwood Park,York University                                      30     30   
Parkwoods                                                           30     30   
Silver Hills,York Mills                                             30     30   
Victoria Village                                                    30     30   
Willowdale South                                                    30     30   
Willowdale West                                                     30     30   
York Mills West                                                     30     30   

                                         

I want to find out how many unique categories can be curated from all the returned venues.

In [121]:
print('There are {} uniques categories.'.format(len(northyork_venues['Venue Category'].unique())))

There are 23 uniques categories.


## 2. Analyzing Each Neighbourhood

In [122]:
# one hot encoding
northyork_onehot = pd.get_dummies(northyork_venues[['Venue Category']], prefix="", prefix_sep="")

# I add neighbourhood column back to dataframe
northyork_onehot['Neighbourhood'] = northyork_venues['Neighbourhood'] 

# then I move neighbourhood column to the first column
fixed_columns = [northyork_onehot.columns[-1]] + list(northyork_onehot.columns[:-1])
northyork_onehot = northyork_onehot[fixed_columns]

northyork_onehot.head()

Neighbourhood  Arts & Crafts Store  Burrito Place  Café  Coffee Shop  \
0  Hillcrest Village                    0              0     0            0   
1  Hillcrest Village                    0              0     1            0   
2  Hillcrest Village                    0              0     0            0   
3  Hillcrest Village                    0              0     0            0   
4  Hillcrest Village                    0              0     0            0   

   Fast Food Restaurant  Grocery Store  Indonesian Restaurant  \
0                     0              0                      0   
1                     0              0                      0   
2                     0              1                      0   
3                     0              0                      0   
4                     0              0                      0   

   Japanese Restaurant  Juice Bar  Korean Restaurant  Movie Theater  \
0                    0          0                  0              0   
1                    0          0                  0              0   
2                    0          0                  0              0   
3                    0          0                  0              1   
4                    0          0                  0              0   

   Pet Store  Plaza  Ramen Restaurant  Restaurant  Sandwich Place  \
0          0      0                 0           0               0   
1          0      0                 0           0               0   
2          0      0                 0           0               0   
3          0      0                 0           0               0   
4          0      0                 1           0               0   

   Seafood Restaurant  Shopping Mall  Sports Bar  Steakhouse  Supermarket  \
0                   1              0           0           0            0   
1                   0              0           0           0            0   
2                   0              0           0           0            0   
3                   0              0           0           0            0   
4                   0              0           0           0            0   

   Sushi Restaurant  Vietnamese Restaurant  
0                 0                      0  
1                 0                      0  
2                 0                      0  
3                 0                      0  
4                 0                      0

In [123]:
# The reulting dataframe
northyork_onehot.shape

(720, 24)

###  I then group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category.

In [124]:
northyork_grouped = northyork_onehot.groupby('Neighbourhood').mean().reset_index()
northyork_grouped

Neighbourhood  Arts & Crafts Store  \
0   Bathurst Manor,Downsview North,Wilson Heights             0.033333   
1                                 Bayview Village             0.033333   
2                Bedford Park,Lawrence Manor East             0.033333   
3                      CFB Toronto,Downsview East             0.033333   
4                                 Don Mills North             0.033333   
5                               Downsview Central             0.033333   
6                             Downsview Northwest             0.033333   
7                                  Downsview West             0.033333   
8                Downsview,North Park,Upwood Park             0.033333   
9                                 Emery,Humberlea             0.033333   
10                     Fairview,Henry Farm,Oriole             0.033333   
11                Flemingdon Park,Don Mills South             0.033333   
12                                      Glencairn             0.033333   
13                              Hillcrest Village             0.033333   
14                                  Humber Summit             0.033333   
15                Lawrence Heights,Lawrence Manor             0.033333   
16                         Newtonbrook,Willowdale             0.033333   
17                 Northwood Park,York University             0.033333   
18                                      Parkwoods             0.033333   
19                        Silver Hills,York Mills             0.033333   
20                               Victoria Village             0.033333   
21                               Willowdale South             0.033333   
22                                Willowdale West             0.033333   
23                                York Mills West             0.033333   

    Burrito Place      Café  Coffee Shop  Fast Food Restaurant  Grocery Store  \
0        0.033333  0.066667          0.1              0.033333       0.033333   
1        0.033333  0.066667          0.1              0.033333       0.033333   
2        0.033333  0.066667          0.1              0.033333       0.033333   
3        0.033333  0.066667          0.1              0.033333       0.033333   
4        0.033333  0.066667          0.1              0.033333       0.033333   
5        0.033333  0.066667          0.1              0.033333       0.033333   
6        0.033333  0.066667          0.1              0.033333       0.033333   
7        0.033333  0.066667          0.1              0.033333       0.033333   
8        0.033333  0.066667          0.1              0.033333       0.033333   
9        0.033333  0.066667          0.1              0.033333       0.033333   
10       0.033333  0.066667          0.1              0.033333       0.033333   
11       0.033333  0.066667          0.1              0.033333       0.033333   
12       0.033333  0.066667          0.1              0.033333       0.033333   
13       0.033333  0.066667          0.1              0.033333       0.033333   
14       0.033333  0.066667          0.1              0.033333       0.033333   
15       0.033333  0.066667          0.1              0.033333       0.033333   
16       0.033333  0.066667          0.1              0.033333       0.033333   
17       0.033333  0.066667          0.1              0.033333       0.033333   
18       0.033333  0.066667          0.1              0.033333       0.033333   
19       0.033333  0.066667          0.1              0.033333       0.033333   
20       0.033333  0.066667          0.1              0.033333       0.033333   
21       0.033333  0.066667          0.1              0.033333       0.033333   
22       0.033333  0.066667          0.1              0.033333       0.033333   
23       0.033333  0.066667          0.1              0.033333       0.033333   

    Indonesian Restaurant  Japanese Restaurant  Juice Bar  Korean Restaurant  \
0                0.033333             0.033333   0.033333           0.066667   

In [125]:
# The new size is:
northyork_grouped.shape

(24, 24)

#### Below I uprint each neighborhood along with the top 3 most common venues.

In [126]:
num_top_venues = 3

for hood in northyork_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = northyork_grouped[northyork_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor,Downsview North,Wilson Heights----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Bayview Village----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Bedford Park,Lawrence Manor East----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----CFB Toronto,Downsview East----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Don Mills North----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Downsview Central----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Downsview Northwest----
              venue  freq
0  Ramen Restaurant  0.13
1       Coffee Shop  0.10
2              Café  0.07


----Downsview West----
   

The following function sorts the venues in descending order.

In [127]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

I then create the new dataframe and display the top 5 venues for each neighbourhood.

In [128]:
num_top_venues = 7

indicators = ['st', 'nd', 'rd']

# I create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# I create a new dataframe
ny_neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
ny_neighbourhoods_venues_sorted['Neighbourhood'] = northyork_grouped['Neighbourhood']

for ind in np.arange(northyork_grouped.shape[0]):
        ny_neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(northyork_grouped.iloc[ind, :], num_top_venues)

ny_neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0  Bathurst Manor,Downsview North,Wilson Heights      Ramen Restaurant   
1                                Bayview Village      Ramen Restaurant   
2               Bedford Park,Lawrence Manor East      Ramen Restaurant   
3                     CFB Toronto,Downsview East      Ramen Restaurant   
4                                Don Mills North      Ramen Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop                  Café     Korean Restaurant   
1           Coffee Shop                  Café     Korean Restaurant   
2           Coffee Shop                  Café     Korean Restaurant   
3           Coffee Shop                  Café     Korean Restaurant   
4           Coffee Shop                  Café     Korean Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  
0  Vietnamese Restaurant         Movie Theater         Burrito Place  
1  Vietnamese Restaurant         Movie Theater         Burrito Place  
2  Vietnamese Restaurant         Movie Theater         Burrito Place  
3  Vietnamese Restaurant         Movie Theater         Burrito Place  
4  Vietnamese Restaurant         Movie Theater         Burrito Place

## 3. Clustering Neighbourhoods

I run k-means to cluster the neighborhood into 4 clusters.

In [129]:
# I set number of clusters
kclusters = 4

northyork_grouped_clustering = northyork_grouped.drop('Neighbourhood', 1)

# I run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(northyork_grouped_clustering)

# I then check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

I create a new dataframe that includes the cluster as well as the top 7 venues for each neighbourhood.

In [130]:
# add clustering labels
ny_neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

northyork_merged = northyork_data

# I merge northyork_merged with northyork_data to add latitude/longitude for each neighbourhood
northyork_merged = northyork_merged.join(ny_neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

northyork_merged.head() # to check the last columns!

Postcode     Borough               Neighbourhood   Latitude  Longitude  \
0      M2H  North York           Hillcrest Village  43.803762 -79.363452   
1      M2J  North York  Fairview,Henry Farm,Oriole  43.778517 -79.346556   
2      M2K  North York             Bayview Village  43.786947 -79.385975   
3      M2L  North York     Silver Hills,York Mills  43.757490 -79.374714   
4      M2M  North York      Newtonbrook,Willowdale  43.789053 -79.408493   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0      Ramen Restaurant           Coffee Shop   
1               0      Ramen Restaurant           Coffee Shop   
2               0      Ramen Restaurant           Coffee Shop   
3               0      Ramen Restaurant           Coffee Shop   
4               0      Ramen Restaurant           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                  Café     Korean Restaurant  Vietnamese Restaurant   
1                  Café     Korean Restaurant  Vietnamese Restaurant   
2                  Café     Korean Restaurant  Vietnamese Restaurant   
3                  Café     Korean Restaurant  Vietnamese Restaurant   
4                  Café     Korean Restaurant  Vietnamese Restaurant   

  6th Most Common Venue 7th Most Common Venue  
0         Movie Theater         Burrito Place  
1         Movie Theater         Burrito Place  
2         Movie Theater         Burrito Place  
3         Movie Theater         Burrito Place  
4         Movie Theater         Burrito Place

### Then we visualize the resulting clusters

In [131]:
# First I create map
map_northyork_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# I add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(northyork_merged['Latitude'], northyork_merged['Longitude'], northyork_merged['Neighbourhood'], northyork_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_northyork_clusters)
       
map_northyork_clusters

## 4. Examining Clusters

Now I examine each cluster and determine the discriminating venue categories that distinguish each cluster.

In [132]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 0, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Borough                                  Neighbourhood   Latitude  \
0   North York                              Hillcrest Village  43.803762   
1   North York                     Fairview,Henry Farm,Oriole  43.778517   
2   North York                                Bayview Village  43.786947   
3   North York                        Silver Hills,York Mills  43.757490   
4   North York                         Newtonbrook,Willowdale  43.789053   
5   North York                               Willowdale South  43.770120   
6   North York                                York Mills West  43.752758   
7   North York                                Willowdale West  43.782736   
8   North York                                      Parkwoods  43.753259   
9   North York                                Don Mills North  43.745906   
10  North York                Flemingdon Park,Don Mills South  43.725900   
11  North York  Bathurst Manor,Downsview North,Wilson Heights  43.754328   
12  North York                 Northwood Park,York University  43.767980   
13  North York                     CFB Toronto,Downsview East  43.737473   
14  North York                                 Downsview West  43.739015   
15  North York                              Downsview Central  43.728496   
16  North York                            Downsview Northwest  43.761631   
17  North York                               Victoria Village  43.725882   
18  North York               Bedford Park,Lawrence Manor East  43.733283   
19  North York                Lawrence Heights,Lawrence Manor  43.718518   
20  North York                                      Glencairn  43.709577   
21  North York               Downsview,North Park,Upwood Park  43.713756   
22  North York                                  Humber Summit  43.756303   
23  North York                                Emery,Humberlea  43.724766   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  -79.363452               0      Ramen Restaurant           Coffee Shop   
1  -79.346556               0      Ramen Restaurant           Coffee Shop   
2  -79.385975               0      Ramen Restaurant           Coffee Shop   
3  -79.374714               0      Ramen Restaurant           Coffee Shop   
4  -79.408493               0      Ramen Restaurant           Coffee Shop   
5  -79.408493               0      Ramen Restaurant           Coffee Shop   
6  -79.400049               0      Ramen Restaurant           Coffee Shop   
7  -79.442259               0      Ramen Restaurant           Coffee Shop   
8  -79.329656               0      Ramen Restaurant           Coffee Shop   
9  -79.352188               0      Ramen Restaurant           Coffee Shop   
10 -79.340923               0      Ramen Restaurant           Coffee Shop   
11 -79.442259               0      Ramen Restaurant           Coffee Shop   
12 -79.487262               0      Ramen Restaurant           Coffee Shop   
13 -79.464763               0      Ramen Restaurant           Coffee Shop   
14 -79.506944               0      Ramen Restaurant           Coffee Shop   
15 -79.495697               0      Ramen Restaurant           Coffee Shop   
16 -79.520999               0      Ramen Restaurant           Coffee Shop   
17 -79.315572               0      Ramen Restaurant           Coffee Shop   
18 -79.419750               0      Ramen Restaurant           Coffee Shop   
19 -79.464763               0      Ramen Restaurant           Coffee Shop   
20 -79.445073               0      Ramen Restaurant           Coffee Shop   
21 -79.490074               0      Ramen Restaurant           Coffee Shop   
22 -79.565963               0      Ramen Restaurant           Coffee Shop   
23 -79.532242               0      Ramen Restaurant           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                   Café     Korean Restaurant  Vietnamese Restaurant   
1                   Café     Korea

### Cluster 2

In [133]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 1, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Neighbourhood, Latitude, Longitude, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue]
Index: []

### Cluster 3

In [134]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 2, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Neighbourhood, Latitude, Longitude, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue]
Index: []

### Cluster 4

In [135]:
northyork_merged.loc[northyork_merged['Cluster Labels'] == 3, northyork_merged.columns[[1] + list(range(2, northyork_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Neighbourhood, Latitude, Longitude, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue]
Index: []

The result shows that ALL neighbourhoods have fallen into the same cluster, i.e., cluster 1. This is no wonder since they are in close proximity and have the venues in common.